### ISSUE TITLE
build failed:  ld: cannot find -lboost_filesystem

### ISSUE THREAD

**Issue by Rjerk** [scylladb/seastar#870]
- os: centos7
- kernel: 3.10.0-862.el7.x86_64
- branch:  seastar-20.05-branch (59f7b32d892191bfae336afcdf6a6d4bd236c183)
- cxx compiler:  g++ 9.3.1

I tried to compile seastar on branch `seastar-20.05-branch`, but got error:

```
➜  seastar git:(seastar-20.05-branch) ✗ ninja -C build/dev/ -j32
ninja: Entering directory `build/dev/'
[65/249] Building CXX object CMakeFiles/seastar.dir/src/net/posix-stack.cc.o
FAILED: CMakeFiles/seastar.dir/src/net/posix-stack.cc.o
/opt/rh/devtoolset-9/root/usr/bin/g++  -DSEASTAR_API_LEVEL=2 -DSEASTAR_HAVE_ASAN_FIBER_SUPPORT -DSEASTAR_HAVE_HWLOC -DSEASTAR_HAVE_LZ4_COMPRESS_DEFAULT -DSEASTAR_HAVE_NUMA -DSEASTAR_THREAD_STACK_GUARDS -DSEASTAR_TYPE_ERASE_MORE -I../../include -Igen/include -I../../src -Igen/src -isystem _cooking/installed/include -O1   -std=gnu++17 -U_FORTIFY_SOURCE -Wno-maybe-uninitialized -DSEASTAR_SSTRING -Wno-error=unused-result -fstack-clash-protection -fvisibility=hidden -UNDEBUG -Wall -Werror -Wno-array-bounds -Wno-error=deprecated-declarations -gz -MD -MT CMakeFiles/seastar.dir/src/net/posix-stack.cc.o -MF CMakeFiles/seastar.dir/src/net/posix-stack.cc.o.d -o CMakeFiles/seastar.dir/src/net/posix-stack.cc.o -c ../../src/net/posix-stack.cc
In file included from ../../src/net/posix-stack.cc:24:
/usr/include/linux/if.h:211:19: error: field ‘ifru_addr’ has incomplete type ‘sockaddr’
  211 |   struct sockaddr ifru_addr;
      |                   ^~~~~~~~~
In file included from /usr/include/linux/if.h:23,
                 from ../../src/net/posix-stack.cc:24:
/usr/include/linux/socket.h:19:27: note: forward declaration of ‘struct sockaddr’
   19 | } __attribute__ ((aligned(_K_SS_ALIGNSIZE))); /* force desired alignment */
      |                           ^~~~~~~~~~~~~~~
In file included from ../../src/net/posix-stack.cc:24:
/usr/include/linux/if.h:212:19: error: field ‘ifru_dstaddr’ has incomplete type ‘sockaddr’
  212 |   struct sockaddr ifru_dstaddr;
      |                   ^~~~~~~~~~~~
In file included from /usr/include/linux/if.h:23,
                 from ../../src/net/posix-stack.cc:24:
/usr/include/linux/socket.h:19:27: note: forward declaration of ‘struct sockaddr’
   19 | } __attribute__ ((aligned(_K_SS_ALIGNSIZE))); /* force desired alignment */
      |                           ^~~~~~~~~~~~~~~
In file included from ../../src/net/posix-stack.cc:24:
/usr/include/linux/if.h:213:19: error: field ‘ifru_broadaddr’ has incomplete type ‘sockaddr’
  213 |   struct sockaddr ifru_broadaddr;
      |                   ^~~~~~~~~~~~~~
In file included from /usr/include/linux/if.h:23,
                 from ../../src/net/posix-stack.cc:24:
/usr/include/linux/socket.h:19:27: note: forward declaration of ‘struct sockaddr’
   19 | } __attribute__ ((aligned(_K_SS_ALIGNSIZE))); /* force desired alignment */
      |                           ^~~~~~~~~~~~~~~
In file included from ../../src/net/posix-stack.cc:24:
/usr/include/linux/if.h:214:19: error: field ‘ifru_netmask’ has incomplete type ‘sockaddr’
  214 |   struct sockaddr ifru_netmask;
      |                   ^~~~~~~~~~~~
In file included from /usr/include/linux/if.h:23,
                 from ../../src/net/posix-stack.cc:24:
/usr/include/linux/socket.h:19:27: note: forward declaration of ‘struct sockaddr’
   19 | } __attribute__ ((aligned(_K_SS_ALIGNSIZE))); /* force desired alignment */
      |                           ^~~~~~~~~~~~~~~
In file included from ../../src/net/posix-stack.cc:24:
/usr/include/linux/if.h:215:20: error: field ‘ifru_hwaddr’ has incomplete type ‘sockaddr’
  215 |   struct  sockaddr ifru_hwaddr;
      |                    ^~~~~~~~~~~
In file included from /usr/include/linux/if.h:23,
                 from ../../src/net/posix-stack.cc:24:
/usr/include/linux/socket.h:19:27: note: forward declaration of ‘struct sockaddr’
   19 | } __attribute__ ((aligned(_K_SS_ALIGNSIZE))); /* force desired alignment */
      |                           ^~~~~~~~~~~~~~~
[96/249] Building CXX object CMakeFiles/seastar.dir/src/core/reactor.cc.o
ninja: build stopped: subcommand failed.

```

after add a line before `#include <linux/if.h>` in `src/net/posix-stack.cc`, the errors gone.

```
// new line
#include <sys/socket.h>

#include <linux/if.h>
```

Maybe somebody can fix it.

The real problem is when I tried to compile it on the master branch using these commands:

```
./configure.py --mode=dev --cook Boost --cook c-ares --cook fmt --cook yaml-cpp --cook hwloc
ninja -C build/dev/ -j32
```

(And I bypassed the failure issued in https://github.com/scylladb/seastar/issues/862)

but still got the error:

```
➜  seastar git:(master) ✗ ninja -C build/dev/ -j32
ninja: Entering directory `build/dev/'
[122/225] Generating other.crt
Signature ok
subject=/C=SE/ST=Stockholm/L=Stockholm/O=apa.org/OU=apa.org/CN=other.apa.org/emailAddress=postmaster@apa.org
Getting CA Private Key
[124/225] Generating test.crt
Signature ok
subject=/C=SE/ST=Stockholm/L=Stockholm/O=scylladb.org/OU=scylladb.org/CN=test.scylladb.org/emailAddress=postmaster@scylladb.org
Getting CA Private Key
[205/225] Linking CXX executable tests/unit/tls_test
FAILED: tests/unit/tls_test
: && /opt/rh/devtoolset-9/root/usr/bin/g++  -O1   tests/unit/CMakeFiles/test_unit_tls.dir/tls_test.cc.o  -o tests/unit/tls_test  -Wl,-rpath,/log/seastar/build/dev/_cooking/installed/lib  -lboost_filesystem  libseastar_testing.a  _cooking/installed/lib/libboost_unit_test_framework.so  libseastar.a  _cooking/installed/lib/libboost_program_options.so  _cooking/installed/lib/libboost_thread.so  _cooking/installed/lib/libboost_chrono.so  _cooking/installed/lib/libboost_system.so  _cooking/installed/lib/libboost_date_time.so  _cooking/installed/lib/libboost_atomic.so  _cooking/installed/lib/libcares.so  /usr/lib64/libcryptopp.so  _cooking/installed/lib64/libfmt.a  /usr/lib64/liblz4.so  -ldl  /usr/lib64/libgnutls.so  -latomic  /usr/lib64/libsctp.so  /usr/lib64/libprotobuf.so  -lrt  _cooking/installed/lib/libyaml-cpp.so  -lpthread  _cooking/installed/lib/libhwloc.so  /usr/lib64/libnuma.so && :
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: cannot find -lboost_filesystem
collect2: error: ld returned 1 exit status
[214/225] Building CXX object apps/memcached/CMakeFiles/app_memcached.dir/memcache.cc.o
ninja: build stopped: subcommand failed.
```

**Comment by nyh** [comment]
@Rjerk is this issue still relevant? 

The missing include file problem was already fixed in commit 1433623962e6abca03dd23ebd1909f9b1a4fce2a, before your report, but I guess was never backported to the branch you were using. But I think instead of a backport we need a new release (@avikivity).

Unfortunately I don't know anything about the "cook" mechanism and I wonder whether any other Seastar developer knows why "--cook Boost" didn't work as expected. It definitely should have built boost_filesystem, but maybe cmake didn't look for it in the right place?

**Comment by Rjerk** [comment]
compiles ok after pulled the latest branch.
